### PR TITLE
[backport] SetN / MapN optimizations

### PR DESCRIPTION
**Pull request by retronym**
Direct implementations of MapN.{filterImpl,exists,forall}

(cherry picked from commit 8821b30ee6202144e3d2cb07c3d27b082ac2574b)

Optimized iterators for immutable.SetN

Avoids projecting to a temporary List.

(cherry picked from commit 3e223d1c7fff68aed51820539085b908c32f6ebf)

Optimized filter for immutable.SetN

Avoids creating iterators and using builders and reuses the
input collection when the predicate selects all elements.

(cherry picked from commit 20b4596bc340ccfc5ba37847d6b21215586f0398)

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -382,7 +382,17 @@ val mimaFilterSettings = Seq {
...
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Set#Set3.filterImpl"),
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Set#Set4.filterImpl"),
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Map#Map2.filterImpl"),
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Map#Map4.filterImpl"),
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Map#Map1.filterImpl"),
+    ProblemFilters.exclude[MissingClassProblem]("scala.collection.immutable.Set$SetNIterator"),
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Set#Set2.filterImpl"),
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Set#Set1.filterImpl"),
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Map#Map3.filterImpl"),
+    )
```
Comment (diesalbla) [inline build.sbt]: To make it correspond to the three commits referred in the description: 

```suggestion
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Map#Map1.filterImpl"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Map#Map2.filterImpl"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Map#Map3.filterImpl"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Map#Map4.filterImpl"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Set#Set1.filterImpl"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Set#Set2.filterImpl"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Set#Set3.filterImpl"),
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Set#Set4.filterImpl"),
    ProblemFilters.exclude[MissingClassProblem]("scala.collection.immutable.Set$SetNIterator"),
  )
```

File: src/library/scala/collection/immutable/Map.scala [hunk 2]
```diff
@@ -153,6 +153,10 @@ object Map extends ImmutableMapFactory[Map] {
     override def foreach[U](f: ((K, V)) => U): Unit = {
       f((key1, value1))
     }
+    override def exists(p: ((K, V)) => Boolean): Boolean = p((key1, value1))
```
Comment (diesalbla) [inline src/library/scala/collection/immutable/Map.scala:156]: I have checked that the changes to this file correspond to those in https://github.com/scala/scala/commit/20b4596bc340ccfc5ba37847d6b21215586f0398#diff-e4f24e9fd33bb1494a43fee25d3b712f

File: src/library/scala/collection/immutable/Set.scala [hunk 3]
```diff
@@ -106,6 +106,29 @@ object Set extends ImmutableSetFactory[Set] {
   }
   private[collection] def emptyInstance: Set[Any] = EmptySet
 
+  @SerialVersionUID(3L)
```
Comment (diesalbla) [inline src/library/scala/collection/immutable/Set.scala:109]: This patch comes from https://github.com/scala/scala/commit/3e223d1c7fff68aed51820539085b908c32f6ebf#diff-e4f24e9fd33bb1494a43fee25d3b712fR130

File: src/library/scala/collection/immutable/Set.scala [hunk 4]
```diff
@@ -164,6 +192,18 @@ object Set extends ImmutableSetFactory[Set] {
     override def forall(@deprecatedName('f) p: A => Boolean): Boolean = {
       p(elem1) && p(elem2)
     }
```
Comment (diesalbla) [inline src/library/scala/collection/immutable/Set.scala:194]: This patch comes from https://github.com/scala/scala/commit/20b4596bc340ccfc5ba37847d6b21215586f0398#diff-e4f24e9fd33bb1494a43fee25d3b712fR209

File: src/library/scala/collection/immutable/Set.scala [hunk 5]
```diff
@@ -201,6 +244,20 @@ object Set extends ImmutableSetFactory[Set] {
     override def forall(@deprecatedName('f) p: A => Boolean): Boolean = {
       p(elem1) && p(elem2) && p(elem3)
     }
+    override private[scala] def filterImpl(pred: A => Boolean, isFlipped: Boolean): Set[A] = {
```
Comment (diesalbla) [inline src/library/scala/collection/immutable/Set.scala:247]: This patch comes from https://github.com/scala/scala/commit/20b4596bc340ccfc5ba37847d6b21215586f0398#diff-e4f24e9fd33bb1494a43fee25d3b712fR260

**Review by diesalbla — APPROVED** [review]

**Review by diesalbla — APPROVED** [review]

**Comment by retronym** [comment]
```
java.nio.file.NoSuchFileException: /tmp/scala.picklecache5122729374671017511/tmp/pipelineBase4056859570065535337/projects/m1/p1/target.jar
```

(known flaky test)

**Pull request merged**